### ISSUE TITLE
[12.0] Unnecessarily long and complex methods on sale wizard

### ISSUE THREAD

**Issue by reinaldoramosarxi** [odoo/odoo#34043]
Hello,

We at our company have been trying to implement Portuguese localization and, as expected, ran into some complications. 

Mostly manageable, of course, however one of these complications involves methods that are very long and complex and will require us to re-implement them whole.

The methods in question are these ones: 
https://github.com/odoo/odoo/blob/12.0/addons/sale/wizard/sale_make_invoice_advance.py#L63
https://github.com/odoo/odoo/blob/12.0/addons/sale/wizard/sale_make_invoice_advance.py#L130

It would help a lot if some changes were made to these methods:

- The _create_invoice method could not require an so_line argument, relying on a mapped() method in the dictionary to fetch the SO lines instead;

- The create dictionary that starts at line 95 could be extracted into a different method;

- The validations that start at line 151 could be extracted into a different method, because the software in Portugal requires the product_id.type to be "other" when it's a down payment;

- The create dictionary that starts at line 164 could also be extracted into a different method.

This would make all implementations that rely on these methods (and localization rely on them a lot) much easier and reliable.

Thank you.

**Comment by Yenthe666** [comment]
Hmm those are rather improvements and they would most likely only happen on the master (upcoming V13) branch.
@tde-banana-odoo or @qdp-odoo could one of you give feedback?

**Comment by tde-banana-odoo** [comment]
@dbo-odoo don't know if anything is planned for this code.

**Comment by dbo-odoo** [comment]
Hi @reinaldoramosarxi,

We can probably do something about it, even though perhaps not all you want:
- splitting dict generating methods in stable is fine, I have no objection there; simply make a PR where you extract these methods as private methods (e.g. `_prepare_line_data` and `_prepare_invoice_data`)
- same for the validation of the product type, it can be extracted safely without any signature change as well (make it private; e.g. `_validate_product`)
  => both these changes should introduce no behavioural change whatsoever and no signature change, just 2 new methods called at the right place
- I'm wayyyy more hesitant about the signature change; even if we do allow for private method signature changes in stable, it is clearly reserved for actual fixes; so this one is a clear no for me in 12.0; besides it's not just a `mapped` call since we need to find the *correct* line (the one with the product selected in the wizard) since only this one will be linked to the down payment invoice line; you can make a PR for master if you want but even then i'm hesitant - this is change for change's sake and while I understand why splitting the dict and validation methods would avoid a complete reimplementation of the methods for inheritance purposes, I fail to see why the line as an argument is problematic

TL;DR:
- PR for 12.0 with splitting of dict and validation methods: OK, simply make a PR and ping me
- Removing `so_line` arg: not in 12.0, possibly in master if you manage to convince me but chances are low

Thx!

KR,

dbo

**Comment by dbo-odoo** [comment]
I'll close this since my last reply seemed sufficient.

**Comment by reinaldoramosarxi** [comment]
Hi,

Sorry, forgot to close it myself.

Thank you!
